File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.java
import java.util.*;
class Solution {
    public int solution(String s) {
        int answer = 0;
        int length = s.length();
        
        for (int i=0;i<length;i++){
            String k = s.substring(i,length) + s.substring(0,i);
            if (solve(k)) answer++;
        }
        
        
        return answer;
    }
    static private boolean solve(String k){
        Stack<Character> stack = new Stack<>();
        
        char[] chars = k.toCharArray();
        
        for (char a : chars) {
            if (a == '(') stack.add(a);
            else if (a == ')') {
                if (stack.isEmpty() || stack.peek() != '(') return false;
                else stack.pop();
            }
            else if (a == '[') stack.add(a);
            else if (a == ']') {
                if (stack.isEmpty() || stack.peek() != '[') return false;
                else stack.pop();
            }
            else if (a == '{') stack.add(a);
            else if (a == '}') {
                if (stack.isEmpty() || stack.peek() != '{') return false;
                else stack.pop();
            }
        }
        if (stack.isEmpty()) return true;
        else return false;
        
    }
}

File: 프로그래머스/2/76502. 괄호 회전하기/괄호 회전하기.cpp
#include <bits/stdc++.h>

using namespace std;
vector<char> v;

bool check(vector<char> & a) {
    stack<char> stack;

    for (char c : a) {
        if (c == '(' || c == '[' || c == '{') {
            stack.push(c);
        } else {
            if (stack.empty()) return false;

            if (c == ')' && stack.top() == '(') stack.pop();
            else if (c == ']' && stack.top() == '[') stack.pop();
            else if (c == '}' && stack.top() == '{') stack.pop();
            else return false;
        }
    }

    return stack.empty();
}

int solution(string s) {
    int answer = 0;

    for (char c : s) {
        v.push_back(c);
    }

    for (int i=0;i<s.size();i++) {
        // 왼쪽으로 i 만큼 회전
        rotate(v.begin(), v.begin() + 1, v.end());
        if (check(v)) answer++; // 괄호 만족하는지 체크
    }

    return answer;
}